File: newsvendor.R
# Newsvendor Model: Umbrella Stocking - Complete R Analysis
# Data collection from Kaggle, cleaning, and optimization
# Pritham Prajwin V | BSD-BG-2512 | ISI Bangalore 2025-26

library(tidyverse)
library(ggplot2)

 
# STEP 1: DATA COLLECTION & LOADING

# Download from: https://www.kaggle.com/datasets/nelgiriyewithana/indian-weather-repository-daily-snapshot/version/231/data
# Alternative: wget/curl the CSV directly
# For this demo, we assume data is downloaded as 'weather_data.csv'

# If downloading via R (requires kaggle CLI setup):
# system('kaggle datasets download -d nelgiriyewithana/indian-weather-repository-daily-snapshot')
# unzip('indian-weather-repository-daily-snapshot.zip')

# Load raw data
weather_raw <- read.csv('weather_data.csv', stringsAsFactors = FALSE)
 
# STEP 2: DATA CLEANING & FILTERING
 

weather <- weather_raw %>%
  filter(location_name == "Bangalore") %>%
  mutate(date = as.Date(last_updated),
         precip_mm = as.numeric(precip_mm),
         month = lubridate::month(date),
         season = case_when(
           month %in% c(7,8,9) ~ "Peak Monsoon",
           month %in% c(6,10,11) ~ "Shoulder",
           TRUE ~ "Dry Season"
         )) %>%
  select(date, precip_mm, month, season) %>%
  arrange(date) %>%
  drop_na(precip_mm) %>%
  filter(precip_mm < 150)

 
# STEP 3: RAINFALL-TO-DEMAND MAPPING (Poisson Model)
 

weather <- weather %>%
  mutate(demand_dist = case_when(
    precip_mm < 2 ~ "Dry",
    precip_mm < 10 ~ "Light",
    precip_mm < 30 ~ "Moderate",
    TRUE ~ "Heavy"
  ),
  lambda = case_when(
    demand_dist == "Dry" ~ 5,
    demand_dist == "Light" ~ 25,
    demand_dist == "Moderate" ~ 55,
    demand_dist == "Heavy" ~ 90
  ),
  demand_est = rpois(n(), lambda))

 
# STEP 4: SEASONAL DEMAND STATISTICS
 

seasonal_stats <- weather %>%
  group_by(season) %>%
  summarise(
    n_days = n(),
    mu = mean(demand_est),
    sigma = sd(demand_est),
    pct_rainy = mean(precip_mm >= 2) * 100,
    .groups = 'drop'
  ) %>%
  arrange(factor(season, levels = c("Peak Monsoon", "Shoulder", "Dry Season")))

 
# STEP 5: NEWSVENDOR OPTIMIZATION
 

c_u <- 70
c_o <- 50
CR <- c_u / (c_u + c_o)

optimal_Q <- function(mu, sigma, CR) {
  round(mu + sigma * qnorm(CR))
}

expected_cost <- function(Q, mu, sigma, c_u, c_o) {
  z <- (Q - mu) / sigma
  std_loss <- dnorm(z) - z * (1 - pnorm(z))
  c_u * sigma * std_loss + c_o * sigma * (z + std_loss)
}

seasonal_results <- seasonal_stats %>%
  mutate(
    Q_star = optimal_Q(mu, sigma, CR),
    opt_cost = mapply(expected_cost, Q_star, mu, sigma, c_u, c_o),
    naive_cost = mapply(expected_cost, 50, mu, sigma, c_u, c_o),
    daily_savings = naive_cost - opt_cost,
    seasonal_savings = daily_savings * n_days
  ) %>%
  select(season, mu, sigma, Q_star, daily_savings, seasonal_savings)

 
# STEP 6: VISUALIZATION - COST CURVE (Peak Monsoon)
 

peak_data <- weather %>% filter(season == "Peak Monsoon")
mu_peak <- mean(peak_data$demand_est)
sigma_peak <- sd(peak_data$demand_est)
Q_star_peak <- optimal_Q(mu_peak, sigma_peak, CR)

Q_range <- seq(10, 150, by = 1)
costs <- sapply(Q_range, function(Q) expected_cost(Q, mu_peak, sigma_peak, c_u, c_o))

p1 <- ggplot(data.frame(Q = Q_range, Cost = costs), aes(x = Q, y = Cost)) +
  geom_line(size = 1.2, color = "#1B2A4A") +
  geom_point(aes(x = Q_star_peak, y = expected_cost(Q_star_peak, mu_peak, sigma_peak, c_u, c_o)),
             size = 4, color = "#C9A84C") +
  geom_vline(xintercept = Q_star_peak, linetype = "dashed", color = "#C9A84C", size = 1) +
  annotate("text", x = Q_star_peak + 5, y = expected_cost(Q_star_peak, mu_peak, sigma_peak, c_u, c_o) - 200,
           label = paste0("Q* = ", Q_star_peak), fontface = "bold", size = 4) +
  labs(title = "Expected Cost vs. Stocking Quantity\n(Peak Monsoon, Jul-Sep)",
       x = "Stocking quantity Q (units)",
       y = "Expected daily cost ₹/C(Q)") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5, size = 12, face = "bold"))

ggsave("plot_cost_curve.png", p1, width = 10, height = 6, dpi = 300)

 
# STEP 7: VISUALIZATION - SENSITIVITY ANALYSIS
 

CR_range <- seq(0.1, 0.95, by = 0.01)
Q_star_range <- mu_peak + sigma_peak * qnorm(CR_range)

p2 <- ggplot(data.frame(CR = CR_range, Q = Q_star_range), aes(x = CR, y = Q)) +
  geom_line(size = 1.2, color = "#1B2A4A") +
  geom_point(aes(x = CR, y = Q_star_peak), size = 4, color = "#C9A84C") +
  geom_vline(xintercept = CR, linetype = "dashed", color = "#C9A84C", size = 1) +
  geom_hline(yintercept = Q_star_peak, linetype = "dotted", color = "#C9A84C", alpha = 0.5) +
  annotate("text", x = CR + 0.05, y = Q_star_peak + 5,
           label = paste0("Our point\nQ*=", Q_star_peak, ", CR=", round(CR, 3)),
           fontface = "bold", size = 3.5) +
  labs(title = "Sensitivity: Q* as a Function of Cost Structure",
       x = "Critical ratio CR = c_u/(c_u + c_o)",
       y = "Optimal Q* (units)") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5, size = 12, face = "bold"))

ggsave("plot_sensitivity.png", p2, width = 10, height = 6, dpi = 300)

 
# STEP 8: FINAL RESULTS
 

print(seasonal_results)
print(paste("Critical Ratio (CR):", round(CR, 4)))

File: newsvendor.py
"""

GROUP-24 : Optimizing Umbrella Inventory for a Street Vendor in Bangalore

"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm

# PARAMETERS
c_u = 70
c_o = 50
CR = c_u / (c_u + c_o)  # Critical ratio = 0.5833

# Peak monsoon demand parameters (from Bangalore rainfall data)
mu_peak = 71.3
sigma_peak = 31.0

# Shoulder and dry season parameters
mu_shoulder = 65.8
sigma_shoulder = 28.5

mu_dry = 15.2
sigma_dry = 7.1

# FUNCTION 1: Calculate Optimal Quantity (Analytical Solution)
def optimal_quantity(mu, sigma, critical_ratio):

    z = norm.ppf(critical_ratio)
    return mu + sigma * z

# FUNCTION 2: Expected Cost for a given Q
def expected_cost(Q, mu, sigma, c_u, c_o):

    z = (Q - mu) / sigma
    std_loss = norm.pdf(z) - z * (1 - norm.cdf(z))

    return c_u * sigma * std_loss + c_o * sigma * (z + std_loss)

# GENERATE DATA FOR PLOT 1: Cost Curve (Peak Monsoon)
Q_values = np.linspace(10, 150, 300)
costs = [expected_cost(Q, mu_peak, sigma_peak, c_u, c_o) for Q in Q_values]

Q_star_peak = optimal_quantity(mu_peak, sigma_peak, CR)
cost_star = expected_cost(Q_star_peak, mu_peak, sigma_peak, c_u, c_o)

# GENERATE DATA FOR PLOT 2: Sensitivity Analysis
CR_values = np.linspace(0.1, 0.95, 200)
Q_star_values = [optimal_quantity(mu_peak, sigma_peak, cr) for cr in CR_values]

# PLOT 1: Cost Curve
plt.figure(figsize=(10, 6))
plt.plot(Q_values, costs, linewidth=2.5, color='#1B2A4A', label='Expected Cost')
plt.plot(Q_star_peak, cost_star, 'o', markersize=12, color='#C9A84C', 
         markeredgecolor='#1B2A4A', markeredgewidth=2, label='Optimal Point')
plt.axvline(x=Q_star_peak, color='#C9A84C', linestyle='--', linewidth=1.5, alpha=0.7)

# Annotations
plt.text(50, 2300, f'steeper\n(₹{c_u}/unit)', fontsize=9, color='#1B2A4A', 
         bbox=dict(boxstyle='round,pad=0.4', facecolor='white', edgecolor='none', alpha=0.8))
plt.text(105, 2050, f'flatter\n(₹{c_o}/unit)', fontsize=9, color='#1B2A4A',
         bbox=dict(boxstyle='round,pad=0.4', facecolor='white', edgecolor='none', alpha=0.8))
plt.text(Q_star_peak + 3, cost_star - 200, f'Q* = {int(Q_star_peak)}', fontsize=10, 
         color='#1B2A4A', bbox=dict(boxstyle='round,pad=0.5', facecolor='#FBF3E2', 
         edgecolor='#C9A84C', linewidth=1.5))

plt.xlabel('Stocking quantity Q (units)', fontsize=11)
plt.ylabel('Expected daily cost ₹/C(Q)', fontsize=11)
plt.title('Expected Cost vs. Stocking Quantity\n(Peak Monsoon, Jul-Sep)', fontsize=12, fontweight='bold')
plt.grid(True, alpha=0.3, linestyle='-', linewidth=0.5)
plt.xlim(10, 150)
plt.ylim(1400, 3400)
plt.tight_layout()
plt.savefig('plot_cost_curve.png', dpi=300, bbox_inches='tight')
plt.close()

print(f"✓ Plot 1 saved: plot_cost_curve.png")
print(f"  Q* (Peak Monsoon) = {Q_star_peak:.1f} units (rounded: {int(Q_star_peak)})")
print(f"  Expected Cost at Q* = ₹{cost_star:.2f}")

# PLOT 2: Sensitivity Analysis
plt.figure(figsize=(10, 6))
plt.plot(CR_values, Q_star_values, linewidth=2.5, color='#1B2A4A')
plt.plot(CR, Q_star_peak, 'o', markersize=12, color='#C9A84C',
         markeredgecolor='#1B2A4A', markeredgewidth=2)

# Add guidelines
plt.axvline(x=CR, color='#C9A84C', linestyle='--', linewidth=1.5, alpha=0.7)
plt.axhline(y=Q_star_peak, color='#C9A84C', linestyle=':', linewidth=1.5, alpha=0.5)
plt.axvline(x=0.5, color='#2980B9', linestyle=':', linewidth=1.5, alpha=0.5, label='symmetric costs')

# Annotations
plt.text(0.5, 20, 'symmetric\ncosts', fontsize=9, color='#2980B9', 
         bbox=dict(boxstyle='round,pad=0.4', facecolor='white', edgecolor='none', alpha=0.8))
plt.text(CR + 0.02, Q_star_peak + 8, f'Our point\nQ* = {int(Q_star_peak)}, CR = {CR:.3f}', 
         fontsize=9, color='#1B2A4A', bbox=dict(boxstyle='round,pad=0.5', 
         facecolor='#FBF3E2', edgecolor='#C9A84C', linewidth=1.5))

plt.xlabel('Critical ratio CR = c_u/(c_u + c_o)', fontsize=11)
plt.ylabel('Optimal Q* (units)', fontsize=11)
plt.title('Sensitivity: Q* as a Function of Cost Structure', fontsize=12, fontweight='bold')
plt.grid(True, alpha=0.3, linestyle='-', linewidth=0.5)
plt.xlim(0.1, 0.95)
plt.ylim(15, 130)
plt.tight_layout()
plt.savefig('plot_sensitivity.png', dpi=300, bbox_inches='tight')
plt.close()

print(f"✓ Plot 2 saved: plot_sensitivity.png")

# SEASONAL ANALYSIS

Q_star_shoulder = optimal_quantity(mu_shoulder, sigma_shoulder, CR)
Q_star_dry = optimal_quantity(mu_dry, sigma_dry, CR)

print(f"\n{'='*60}")
print(f"SEASONAL OPTIMAL STOCKING QUANTITIES")
print(f"{'='*60}")
print(f"Peak Monsoon (Jul-Sep):    Q* = {int(Q_star_peak)} units")
print(f"Shoulder Months (Oct-Jun): Q* = {int(Q_star_shoulder)} units")
print(f"Dry Season (Apr-Jun):      Q* = {int(Q_star_dry)} units")
print(f"\nFormula: Q* = μ + σ × Φ^(-1)(CR), where CR = {CR:.4f}")
